sky/provision/instance_setup.py

```diff
@@ -123,24 +124,20 @@ def _parallel_ssh_with_cache(func,
         max_workers = subprocess_utils.get_parallel_threads()
     with futures.ThreadPoolExecutor(max_workers=max_workers) as pool:
         results = []
-        for instance_id, metadatas in cluster_info.instances.items():
-            for i, metadata in enumerate(metadatas):
-                cache_id = f'{instance_id}-{i}'
-                runner = command_runner.SSHCommandRunner(
-                    metadata.get_feasible_ip(),
-                    port=metadata.ssh_port,
-                    **ssh_credentials)
-                wrapper = metadata_utils.cache_func(cluster_name, cache_id,
-                                                    stage_name, digest)
-                if (cluster_info.head_instance_id == instance_id and i == 0):
-                    # Log the head node's output to the provision.log
-                    log_path_abs = str(provision_logging.get_log_path())
-                else:
-                    log_dir_abs = metadata_utils.get_instance_log_dir(
-                        cluster_name, cache_id)
-                    log_path_abs = str(log_dir_abs / (stage_name + '.log'))
-                results.append(
-                    pool.submit(wrapper(func), runner, metadata, log_path_abs))
+        runners = provision.get_command_runners(cluster_info.provider_name,
+                                                cluster_info, **ssh_credentials)
+        for i, runner in enumerate(runners):
+            cache_id = str(i)
```

Comment:
Should we still use the `f'{instance_id}-{i}'`? IIUC this helps reduce the conflict of the cache

Reply:
Good catch! Updated the implementation to have `{instance_id}-{i}` here.